crowbar_framework/spec/models/api/upgrade_spec.rb

```diff
@@ -367,7 +379,175 @@
+      allow(Node).to(
+        receive(:find).with("state:crowbar_upgrade").
+        and_return([Node.find_by_name("testing.crowbar.com")])
+      )
+      allow(Api::Upgrade).to receive(:join_ceph_nodes).and_return(true)
+
+      allow_any_instance_of(Crowbar::UpgradeStatus).to receive(:progress).and_return(
+        suggested_upgrade_mode: :normal
+      )
+      expect(Crowbar::UpgradeStatus.new.upgrade_mode).to be :normal
+
+      allow(BarclampCatalog).to receive(:barclamps).and_return(barclamp_catalog)
+      allow(BarclampCatalog).to receive(:category).and_return("OpenStack")
+      allow(BarclampCatalog).to receive(:category).with("deployer").and_return("Crowbar")
+
+      cinder_proposal.elements["cinder-controller"] = ["drbd"]
+
+      nova_proposal.elements["nova-controller"] = ["ceph"]
+      nova_proposal.elements["nova-compute-kvm"] = ["testing"]
+
+      allow(Proposal).to receive(:where).with(barclamp: "database").and_return([database_proposal])
+      allow(Proposal).to receive(:where).with(barclamp: "cinder").and_return([cinder_proposal])
+      allow(Proposal).to receive(:where).with(barclamp: "nova").and_return([nova_proposal])
+      allow_any_instance_of(Proposal).to(receive(:active?).and_return(true))
+
+      allow(BarclampCatalog).to receive(:run_order).with("database").and_return(1)
+      allow(BarclampCatalog).to receive(:run_order).with("cinder").and_return(2)
+      allow(BarclampCatalog).to receive(:run_order).with("nova").and_return(3)
+
+      allow(Api::Upgrade).to receive(:upgrade_one_node).and_return(true)
+
+      # rest of the upgrade, after controller nodes
+      allow(Api::Upgrade).to receive(:prepare_all_compute_nodes).and_return(true)
+      allow(Api::Upgrade).to receive(:upgrade_all_compute_nodes).and_return(true)
+      allow(Api::Upgrade).to receive(:finalize_nodes_upgrade).and_return(true)
+      allow_any_instance_of(Crowbar::UpgradeStatus).to receive(
+        :start_step
+      ).with(:nodes).and_return(true)
+      allow_any_instance_of(Crowbar::UpgradeStatus).to receive(:end_step).and_return(true)
+
+      expect(subject.class.nodes_without_delay).to be true
+    end
+
+    it "successfully upgrades compute nodes" do
+      allow(Node).to(
+        receive(:find).with("state:crowbar_upgrade").
+        and_return([Node.find_by_name("testing.crowbar.com")])
+      )
+      allow(Api::Upgrade).to receive(:join_ceph_nodes).and_return(true)
+
+      allow_any_instance_of(Crowbar::UpgradeStatus).to receive(:progress).and_return(
+        suggested_upgrade_mode: :normal
+      )
+      expect(Crowbar::UpgradeStatus.new.upgrade_mode).to be :normal
+
+      allow(Api::Upgrade).to receive(:upgrade_controllers_disruptive).and_return(true)
+      allow(Api::Upgrade).to receive(:prepare_all_compute_nodes).and_return(true)
+
+      allow(Node).to(
+        receive(:find).with("roles:nova-compute-kvm").
+        and_return([Node.find_by_name("testing.crowbar.com")])
```

Comment:
Would be good to really test this with multiple compute nodes. And verify that ```execute_scripts_and_wait_for_finish``` is actually called with multiple nodes passed.